plugins/microservices/msi_atomic_apply_metadata_operations/libmsi_atomic_apply_metadata_operations.cpp

```diff
@@ -0,0 +1,74 @@
+#include "irods_ms_plugin.hpp"
+#include "irods_re_structs.hpp"
+#include "msParam.h"
+#include "rodsErrorTable.h"
+#include "rs_atomic_apply_metadata_operations.hpp"
+#include "irods_error.hpp"
+#include "rodsLog.h"
+
+#include <functional>
+#include <string>
+#include <exception>
+
+namespace
+{
+    auto to_string(msParam_t& _p) -> const char*
+    {
+        const auto* s = parseMspForStr(&_p);
+
+        if (!s) {
+            THROW(SYS_INVALID_INPUT_PARAM, "Failed to convert microservice argument to string.");
+        }
+
+        return s;
+    }
+
+    auto msi_impl(msParam_t* _json_input, msParam_t* _json_output, ruleExecInfo_t* _rei) -> int
+    {
+        if (!_json_input || !_json_output) {
+            rodsLog(LOG_ERROR, "Invalid input argument.");
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        try {
+            const auto* json_input = to_string(*_json_input);
+            char* json_output{};
+
+            if (const auto ec = rs_atomic_apply_metadata_operations(_rei->rsComm, json_input, &json_output); ec != 0) {
```

Comment:
should be invoking the `server_api_call` - this will go back through the pep machinery and allow PEPs to fire around this operation.

Reply:
This call is actually a convenience wrapper around server_api_call. I've verified (at the bench) that the PEPs are firing whenever the API plugin is invoked.

I'm now updating the tests to do this verification as well.

Reply:
test now checks that PEPs fire when they are supposed to.